### PR TITLE
[AIROCMLIR-448] Lower `linalg.generic` Backwards Conv into `rock.bwd_data_conv`

### PR DESCRIPTION
**Pull request by Mr-Anyone**
## Motivation

Being able to lower a `linalg.generic` into `rock.bwd_data_conv`. 

## Technical Details

Essentially, we are converting a `linalg.generic` with a NGC*, GFC*, and NGF* into its equivalent form in `rock.bwd_data_conv`. There are only two special cases: backward convolution with padding, and backward convolution with dilation and strides which case the result tensor to have empty values. 

If we have padding, we first remove the padding from the code structure. 
```mlir
%backward_conv = linalg.generic ins(...) outs(...)  // compute backward convolution
%collapsed_group = tensor.collapse_shape %backward_conv // collapsing group dimension
%handle_padding= tensor.extract_slice %collapsed_group // handle padding
```

to 

```mlir
%1 = rock.conv_bwd_data(%expanded_2, %0, %expanded_1) .... // compute backwards convolution
%collapsed = tensor.collapse_shape %0 [[0], [1, 2], [3], [4]] : // collapsing group dimensions
```
Here, we have remove the `tensor.extract_slice`, which is the padding from the linalg pipeline as `rock.conv_bwd_data` handles that for us. 

In the cases that we need to prefill the result tensor, we emit a rock.prefill attribute like the one found in tosa pipeline. As of current, **this is hard coded** as we assume the result tensor will always be at the 0th index. 

## Test Plan

Added e2e test for linalg to rock. 

## Test Result

Pass all e2e test.

## Submission Checklist

- [X] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

File: mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp [hunk 1]
```diff
@@ -323,6 +333,122 @@ ConvLinalgConverter::isConv(ConversionPatternRewriter &rewriter,
...
 }
 
+LogicalResult BwdConvLinalgConverter::matchAndRewrite(
+    linalg::GenericOp op, OpAdaptor adaptor,
+    ConversionPatternRewriter &rewriter) const {
+  FailureOr<ConvFields> maybeConv = isConv(rewriter, op);
+  if (failed(maybeConv))
+    return failure();
+
+  ConvFields conv = *maybeConv;
```
Comment (umangyadav) [inline mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp:476]: can you add lit tests ?

File: mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp [hunk 2]
```diff
@@ -323,6 +333,122 @@ ConvLinalgConverter::isConv(ConversionPatternRewriter &rewriter,
...
+  case rock::LinalgConvType::Conv3dBWDNgchwdGcfhwd:
+    break;
+  default:
+    return failure();
+  }
+  bool hasPadding = llvm::any_of(conv.padding, [](Attribute attr) {
+    return cast<IntegerAttr>(attr).getInt() != 0;
+  });
+
+  // FIXME: add a check here!
```
Comment (umangyadav) [inline mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp]: can you add this checks ?

File: mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp [hunk 3]
```diff
@@ -323,6 +333,122 @@ ConvLinalgConverter::isConv(ConversionPatternRewriter &rewriter,
...
+  auto dilationDims = ctx.getDilationVal();
+  auto filterDims = ctx.getConvDims().fil;
+  auto numKernels =
+      rock::backwardDataKernelIds(strideDims, dilationDims, filterDims,
+                                  /*usesV4R1=*/true);
+  // If there is no zeroinit kernel needed, then there is nothing more we need
+  // to do here.
+  if (!rock::isEveryElementWrittenBwdData(strideDims, dilationDims,
+                                          filterDims)) {
+    // FIXME: don't hard code this - see PR#1687
```
Comment (umangyadav) [inline mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp]: is this FIXME relevant ?

File: mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp (hunk 2 above)
Comment (Mr-Anyone) [reply mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp]: The checks was that I have to remove the padding, which I have just done.

File: mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp (hunk 3 above)
Comment (Mr-Anyone) [reply mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp]: Currently, I have hard coded this to the first output with `Type funcResType=func.getFunctionType().getResult(0);`

My understanding of mlir is that there can be multiple output. In those cases, we have to climb a tree to figure which of the output we need to set rock.prefill attribute?

File: mlir/test/fusion/pr-e2e/mixr-bwd-data-conv-padding1.mlir [hunk 4]
```diff
@@ -1,5 +1,6 @@
...
+// RUN: rocmlir-gen -fut mlir_bwd_data_conv --arch %arch --clone-harness %s | rocmlir-driver -kernel-pipeline=migraphx-linalg,highlevel -host-pipeline=migraphx,highlevel -targets %arch | rocmlir-gen -ph -rand 1 -rand_type float -fut mlir_bwd_data_conv_wrapper --verifier clone -relDiff_threshold 0.00001 - | rocmlir-driver -host-pipeline mhal,runner -kernel-pipeline full -targets %arch  | xmir-runner --shared-libs=%linalg_test_lib_dir/libmlir_rocm_runtime%shlibext,%conv_validation_wrapper_library_dir/libconv-validation-wrappers%shlibext,%linalg_test_lib_dir/libmlir_runner_utils%shlibext,%linalg_test_lib_dir/libmlir_float16_utils%shlibext,%linalg_test_lib_dir/libmlir_c_runner_utils%shlibext,%linalg_test_lib_dir/libmlir_async_runtime%shlibext --entry-point-result=void | FileCheck %s
```
Comment (umangyadav) [inline mlir/test/fusion/pr-e2e/mixr-bwd-data-conv-padding1.mlir]: I see some inconsistencies

Some test have `host-pipeline=migraphx-linalg,highlevel `
and some tests have `host-pipeline=migraphx,highlevel`

File: mlir/test/Conversion/LinalgToRock/linalg-to-rock-bwd-conv-2d.mlir [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    ^bb0(%in: f32, %in_2: f32, %out: f32):
+        %1 = arith.mulf %in, %in_2 : f32
+        %2 = arith.addf %out, %1 : f32
+        linalg.yield %2 : f32
+    } -> tensor<1x1x1x5x5xf32>
+    %collapsed = tensor.collapse_shape %0 [[0], [1, 2], [3], [4]] : tensor<1x1x1x5x5xf32> into tensor<1x1x5x5xf32>
+    %extracted_slice = tensor.extract_slice %collapsed[0, 0, 1, 1] [1, 1, 3, 3] [1, 1, 1, 1] : tensor<1x1x5x5xf32> to tensor<1x1x3x3xf32>
+    %collapsed_1 = tensor.collapse_shape %extracted_slice [[0, 1, 2, 3]] : tensor<1x1x3x3xf32> into tensor<9xf32>
+    return %collapsed_1 : tensor<9xf32>
+}
```
Comment (umangyadav) [inline mlir/test/Conversion/LinalgToRock/linalg-to-rock-bwd-conv-2d.mlir]: Nit: new line

File: mlir/test/Conversion/LinalgToRock/linalg-to-rock-bwd-conv-2d.mlir [hunk 6]
```diff
@@ -0,0 +1,38 @@
+// RUN: sed s/##TOKEN_ARCH##/%arch/g %s | rocmlir-opt --linalg-to-rock -verify-diagnostics --split-input-file | FileCheck %s
+
+// Output: NGCHW = 1x1x1x3x3, Filter: GCKHW = 1x1x1x3x3
+// stride=1, dilation=1, padding=[1,1,1,1], group=1
```
Comment (umangyadav) [inline mlir/test/Conversion/LinalgToRock/linalg-to-rock-bwd-conv-2d.mlir]: Nit:

Would be good to add some more tests with group != 1 
With 3D conv_Bwd

with different strides,dilations and padding values

File: mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp [hunk 7]
```diff
@@ -323,6 +333,122 @@ ConvLinalgConverter::isConv(ConversionPatternRewriter &rewriter,
...
+    Type funcResType = func.getFunctionType().getResult(0);
+    auto shapedResType = cast<ShapedType>(funcResType);
+    Type elementType = shapedResType.getElementType();
+    if (isa<FloatType>(elementType)) {
+      outputInitVal = rewriter.getFloatAttr(elementType, 0.0);
+    } else if (isa<IntegerType>(elementType)) {
+      outputInitVal = rewriter.getIntegerAttr(elementType, 0);
+    } else {
+      // We only expect integer and float types for now
+      assert(false && "Unsupported element type for prefill attribute");
```
[the remark this answers is not in the document]
Comment (umangyadav) [reply mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp]: @Mr-Anyone  i think this is a good suggestion

File: mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp (hunk 3 above)
Comment (umangyadav) [reply mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp]: > My understanding of mlir is that there can be multiple output. In those cases, we have to climb a tree to figure which of the output we need to set rock.prefill attribute? 

Yes there can be multiple output and you need to trace to find out which one is result of  the conv_bwd

File: mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp [hunk 8]
```diff
@@ -323,6 +333,122 @@ ConvLinalgConverter::isConv(ConversionPatternRewriter &rewriter,
...
+  auto output =
+      bufferization::AllocTensorOp::create(rewriter, loc, resultShape, {});
+  auto cop = rock::ConvBwdDataOp::create(
+      rewriter, loc, output.getType(), filter, output, input,
+      /*features=*/nullptr,
+      /*blockSize=*/nullptr,
+      /*gridSize=*/nullptr, conv.padding, conv.stride, conv.dilation,
+      /*params=*/nullptr, rewriter.getIndexAttr(0),
+      /*usesV4R1=*/rewriter.getBoolAttr(false));
+  setConvLayoutAttrs(rewriter, cop, getSpatialDim(conv.type));
```
[the remark this answers is not in the document]
Comment (umangyadav) [reply mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp:538]: @Mr-Anyone  can you check this ?

File: mlir/test/Conversion/LinalgToRock/linalg-to-rock-bwd-conv-2d.mlir [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+#map = affine_map<(d0, d1, d2, d3, d4, d5, d6, d7) -> (d0, d1, d5, d2, d3)>
```
Comment (umangyadav) [inline mlir/test/Conversion/LinalgToRock/linalg-to-rock-bwd-conv-2d.mlir]: Can you add tests to cover negative lines ?

File: mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp [hunk 10]
```diff
@@ -323,6 +333,122 @@ ConvLinalgConverter::isConv(ConversionPatternRewriter &rewriter,
...
+    tensor::CollapseShapeOp collapseGroupDim =
+        tensor::CollapseShapeOp::create(rewriter, loc, output, reassocations);
+    rewriter.eraseOp(op);
+    rewriter.eraseOp(collapseGroupPadding);
+    rewriter.replaceOp(extractSlicePadding, collapseGroupDim);
+    return success();
+  }
+
+  rewriter.replaceOp(op, output);
+  return success();
```
[the remark this answers is not in the document]
Comment (umangyadav) [reply mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp:588]: @Mr-Anyone  i think this is a better way to replaceOp. But consider this as a "nit"

File: mlir/test/Conversion/LinalgToRock/linalg-to-rock-bwd-conv-2d.mlir [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+// CHECK-SAME: strides = [1 : index, 1 : index]
+// CHECK-DAG: %[[collapsed:.*]] = tensor.collapse_shape %[[alloc]]
+// CHECK-DAG: %[[collapsed_1:.*]] = tensor.collapse_shape %[[collapsed]]
+// CHECK-DAG: return %[[collapsed_1]]
+func.func @mlir_bwd_data_conv(%arg0: tensor<9xf32>, %arg1: tensor<9xf32>) -> tensor<9xf32> attributes {arch = "##TOKEN_ARCH##", kernel} {
+    %cst = arith.constant dense<0.000000e+00> : tensor<1x1x1x5x5xf32>
+    %expanded = tensor.expand_shape %arg1 [[0, 1, 2, 3, 4]] output_shape [1, 1, 1, 3, 3] : tensor<9xf32> into tensor<1x1x1x3x3xf32>
+    %expanded_0 = tensor.expand_shape %arg0 [[0, 1, 2, 3, 4]] output_shape [1, 1, 1, 3, 3] : tensor<9xf32> into tensor<1x1x1x3x3xf32>
+    %0 = linalg.generic {indexing_maps = [#map, #map1, #map2], iterator_types = ["parallel", "parallel", "parallel", "parallel", "parallel", "reduction", "reduction", "reduction"]} ins(%expanded, %expanded_0 : tensor<1x1x1x3x3xf32>, tensor<1x1x1x3x3xf32>) outs(%cst : tensor<1x1x1x5x5xf32>) attrs =  {conv_op = #rock<LinalgConvType convbwd2d_ngchw_gckhw>, dilation = [1, 1], group = 1 : i64, pad = [1, 1, 1, 1], stride = [1, 1]} {
```
Comment (umangyadav) [inline mlir/test/Conversion/LinalgToRock/linalg-to-rock-bwd-conv-2d.mlir]: add test with `perf_config` to make sure it is propagated ?

File: mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (Mr-Anyone) [reply mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp:538]: This is correct in a sense that BWDConv1D doesn't work, but the fix is wrong. The fundamental reasons is that I think `rock.conv` only support cases where spatial dim > 1?

Currently, we do the expansion for Conv1D in the Linalg -> Rock lowering. I think the better fix is to move this expansion logic into MIGraphX -> Linalg, and let rock-view-to-transforms convert the tensor operations into rock.transform.

File: mlir/test/Conversion/LinalgToRock/linalg-to-rock-bwd-conv-2d.mlir (hunk 11 above)
Comment (Mr-Anyone) [reply mlir/test/Conversion/LinalgToRock/linalg-to-rock-bwd-conv-2d.mlir]: I have added the perf_config attribute to this test as well.

File: mlir/test/Conversion/LinalgToRock/linalg-to-rock-bwd-conv-2d.mlir (hunk 6 above)
Comment (Mr-Anyone) [reply mlir/test/Conversion/LinalgToRock/linalg-to-rock-bwd-conv-2d.mlir]: I added a group 3D backward convolution e2e test with  mixr-bwd-data-conv-3d-group with non standard stride / dilation/ padding.

File: mlir/test/fusion/pr-e2e/mixr-bwd-data-conv-padding1.mlir (hunk 4 above)
Comment (Mr-Anyone) [reply mlir/test/fusion/pr-e2e/mixr-bwd-data-conv-padding1.mlir]: I have set everything int --host-pipeline=migraphx-linalg,highlevel and --kernel-pipeline=migraphx-linalg,highlevel.

File: mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp (hunk 3 above)
Comment (Mr-Anyone) [reply mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp]: I have added a test to see if there are multiple return values. 

If there is only one return value, the returnIndex must be 0.

File: mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp [hunk 12]
```diff
@@ -456,6 +466,127 @@ ConvLinalgConverter::isConv(ConversionPatternRewriter &rewriter,
...
+           "these op should have be found from before");
+    SmallVector<ReassociationIndices, 4> reassocations{{0}, {1, 2}};
+    llvm::transform(llvm::seq<int64_t>(3, 3 + conv.spatialDim),
+                    std::back_inserter(reassocations),
+                    [](int64_t index) { return ReassociationIndices{index}; });
+    tensor::CollapseShapeOp collapseGroupDim =
+        tensor::CollapseShapeOp::create(rewriter, loc, output, reassocations);
+    rewriter.eraseOp(op);
+    rewriter.eraseOp(collapseGroupPadding);
+    rewriter.replaceOp(extractSlicePadding, collapseGroupDim);
```
Comment (justinrosner) [inline mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp:583]: Can we erase op when it's result is still consumed by collapseGroupPadding?

File: mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp [hunk 13]
```diff
@@ -456,6 +466,127 @@ ConvLinalgConverter::isConv(ConversionPatternRewriter &rewriter,
...
+  auto dilationDims = ctx.getDilationVal();
+  auto filterDims = ctx.getConvDims().fil;
+  // If there is no zeroinit kernel needed, then there is nothing more we need
+  // to do here.
+  // FIXME: don't hard code this - see PR#1687 See AIROCMLIR-748 for more
+  // details
+  if (!rock::isEveryElementWrittenBwdData(strideDims, dilationDims,
+                                          filterDims)) {
+    func::FuncOp func = op->getParentOfType<func::FuncOp>();
+    if (func.getResultTypes().size() > 1) {
```
Comment (justinrosner) [inline mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp]: Nit: Shouldn't we be checking for exactly one result here? Since this is run before bufferization takes place?

File: mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp [hunk 14]
```diff
@@ -456,6 +466,127 @@ ConvLinalgConverter::isConv(ConversionPatternRewriter &rewriter,
...
+    if (isa<FloatType>(elementType)) {
+      outputInitVal = rewriter.getFloatAttr(elementType, 0.0);
+    } else if (isa<IntegerType>(elementType)) {
+      outputInitVal = rewriter.getIntegerAttr(elementType, 0);
+    } else {
+      // We only expect integer and float types for now
+      return op.emitError("unsupported element type for prefill attribute");
+    }
+
+    func.setResultAttr(0, rock::PrefillAttr::getMnemonic(), outputInitVal);
```
Comment (justinrosner) [inline mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp:569]: In the Tosa pipeline don't we have tracing logic to find which arg to set the prefill attribute on? Can we reuse that logic here?

File: mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp [hunk 15]
```diff
@@ -456,6 +466,127 @@ ConvLinalgConverter::isConv(ConversionPatternRewriter &rewriter,
...
+  }
+
+  if (hasPadding) {
+    assert(extractSlicePadding && collapseGroupPadding &&
+           "these op should have be found from before");
```
Comment (justinrosner) [inline mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp]: Nit: `op` -> `ops`

File: mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp (hunk 13 above)
Comment (Mr-Anyone) [reply mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp]: I have changed it to be `!= 1` instead since I realized it could be 0 in some cases.

File: mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp (hunk 14 above)
Comment (Mr-Anyone) [reply mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp:569]: We could, but I think you would have to link against TosaToRock. 

I think the same goes from setSplitK, and Gemm+Geem matching logic. It is general enough that it basically works out of the box with small changes.

File: mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp (hunk 12 above)
Comment (Mr-Anyone) [reply mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp:583]: I would assume so because [erase](https://mlir.llvm.org/docs/DialectConversion/#immediate-vs-delayed-ir-modification) is delayed? 

But we have to replace the extract_slice the rock.bwd_data_conv because the extract_slice has is the padded output. Transposed convolution work in the [opposite way](https://d2l.ai/chapter_computer-vision/transposed-conv.html#padding-strides-and-multiple-channels).

File: mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (umangyadav) [reply mlir/lib/Conversion/LinalgToRock/LinalgToRock.cpp:538]: Can you open a ticket for better fix ? and if you have time and if you want, you can work on it too.

**Review by umangyadav — APPROVED** [review]
